src/collapse/ghcprof.rs

```
@@ -0,0 +1,240 @@
+    Bytes,
+}
+
+/// A stack collapser for the output of `ghc`'s prof files.
+///
+/// To construct one, either use `ghcprof::Folder::default()` or create an [`Options`] and use
+/// `ghcprof::Folder::from(options)`.
+#[derive(Clone, Default)]
+pub struct Folder {
+    /// Cost for the current stack frame.
+    current_cost: usize,
+
+    /// Function on the stack in this entry thus far.
+    stack: Vec<String>,
+
+    opt: Options,
+}
+
+// The starting character offset of important columns
+#[derive(Debug)]
+struct Cols {
+    cost_centre: usize,
+    module: usize,
+    source: usize,
+}
+
+impl Collapse for Folder {
+    fn collapse<R, W>(&mut self, mut reader: R, writer: W) -> io::Result<()>
+    where
+        R: io::BufRead,
+        W: io::Write,
+    {
+        // Consume the header...
+        let mut line = Vec::new();
+        let cols = loop {
+            line.clear();
+            if reader.read_until(b'\n', &mut line)? == 0 {
+                warn!("File ended before start of call graph");
+                return Ok(());
+            };
+            let l = String::from_utf8_lossy(&line);
+
+            if l.split_whitespace()
+                .take(START_LINE.len())
+                .eq(START_LINE.iter().cloned())
+            {
+                let cost_centre = 0;
+                let module = l.find("MODULE").unwrap_or(0);
+                // Pick out these fixed columns, first two are individual only
+                // "%time %alloc   %time %alloc"
+                // `ticks` and `bytes` columns are optional and might appear on the end
+                // ticks header is right aligned
+                // bytes header is right aligned
+                //   - BUT it has a max width of 9 whilst its values can exceed (but are always space separted)
+                // "%time %alloc   %time %alloc  ticks  bytes"
+                let source = match self.opt.source {
+                    Source::PercentTime => l.find("%time").unwrap(),
```

This still panics, but should return an error instead (like the other two now do).

---

it's one of those cases where we "know" it's there since otherwise we couldn't be in the block because we just matched it as part of `START_LINE`. I've added a nice return error anyway and made clippy happy

---

Ah, in that case you can use `expect` instead and give the rationale!

---

okay swapped it